backend/src/main/repository/PromotionRepository.ts

```diff
@@ -16,66 +11,111 @@ export class PromotionRepository extends Repository<Promotion> {
+      queryBuilder.andWhere('discount.type = :type', {
         type: promotionQuery.discountType,
       });
     }
 
     if (promotionQuery?.expirationDate) {
-      queryBuilder = queryBuilder.andWhere(
-        'promotion.expirationDate >= :date',
-        {
-          date: new Date(promotionQuery.expirationDate),
-        }
-      );
+      queryBuilder.andWhere('promotion.expirationDate >= :date', {
+        date: new Date(promotionQuery.expirationDate),
+      });
     }
 
-    if (promotionQuery?.name) {
-      let name = promotionQuery.name;
+    if (promotionQuery?.searchQuery) {
+      return this.fullTextSearch(queryBuilder, promotionQuery);
+    }
 
-      // first escape % and _ (request needs to show it as https://stackoverflow.com/questions/17342671/pass-a-percent-sign-in-a-url-and-get-exact-value-of-it-using-php)
-      // this way users can query promotions that contain % or _
-      name = name.replace('%', '\\%');
-      name = name.replace('_', '\\_');
+    return queryBuilder.getMany();
+  }
 
-      // add wildcards
-      name = `%${name.replace(/ /g, '%')}%`;
+  private async fullTextSearch(
+    queryBuilder: SelectQueryBuilder<Promotion>,
+    promotionQuery: PromotionQueryDTO
+  ): Promise<any> {
+    // todo: modify searchQuery accordingly
+    // todo: decide how to handle special characters (e.g. single quotes, AT&T as a single word)
+    const searchQuery = promotionQuery.searchQuery;
 
-      queryBuilder = queryBuilder.andWhere('promotion.name ilike :name', {
-        name: name,
-      });
+    const arrayOfIdRank: IdRank[] = await this.createQueryBuilder('promotion')
+      .select('promotion.id', 'id')
+      .addSelect(
+        `ts_rank_cd(tsvector, replace(plainto_tsquery('${searchQuery}')::text, '&', '|')::tsquery)`,
+        'rank'
+      )
+      .where(
+        `tsvector @@ replace(plainto_tsquery('${searchQuery}')::text, '&', '|')::tsquery`
+      )
+      // .addSelect(`ts_rank_cd(tsvector, plainto_tsquery('${searchQuery}'))`, 'rank')
+      // .where(`tsvector @@ plainto_tsquery('${searchQuery}')`)
+      .orderBy('rank', 'DESC')
+      .getRawMany();
+
+    if (!arrayOfIdRank?.length) {
+      return [];
+    }
+
+    const promotions: Promotion[] = await queryBuilder
+      .innerJoinAndSelect('promotion.discount', 'discount')
+      .andWhere('promotion.id IN (:...ids)', {
+        ids: arrayOfIdRank.map((idRank) => idRank.id),
+      })
+      .getMany();
+
+    if (!promotions?.length) {
+      return [];
     }
 
-    return queryBuilder;
+    const mapIdToPromotion: Map<string, Promotion> = new Map(
+      promotions.map((promotion) => [promotion.id, promotion])
+    );
+
+    const result = [];
+    for (const idRank of arrayOfIdRank) {
+      const promotion = mapIdToPromotion.get(idRank.id);
+      if (promotion) {
+        promotion.rank = idRank.rank;
+        result.push(promotion);
+      }
+    }
+
+    return result;
   }
 }
```

Comment:
This is still a rough algorithm, but here is what's happening
1. `promotionQuery.searchQuery` will contain the keywords we want to search up. (e.g. "Vancouver Chinese dumplings")
2. `arrayOfIdRank` will break down the search query into a tsvector consisted of lexemes and the search query will be using `OR` operators.
     - it will then return a promotionId along with the rank, where the rank denotes how well the promotion name/description maps to the search query
3. `promotions` will contain the result of finishing the rest of the query builder (which may contain `where` clauses for discount/category/expirationdate/etc. and joining with discount, but it will only take promotions that have id inside `arrayOfIdRank`.
4. We add `rank` value to each promotion and return.

Let me know if any of this needs further clarification, I'd be more than happy to discuss!

Reply:
Ignore: will explain this in other issue that concerns fully implementing search query